modules/component/src/core/cd-aware/cd-aware_creator.ts

```diff
@@ -0,0 +1,98 @@
+import {
+  combineLatest,
+  EMPTY,
+  NEVER,
+  NextObserver,
+  Observable,
+  PartialObserver,
+  Subject,
+  Subscribable,
+  Subscription,
+} from 'rxjs';
+import {
+  catchError,
+  distinctUntilChanged,
+  map,
+  startWith,
+  switchMap,
+  tap,
+} from 'rxjs/operators';
+import {
+  CdStrategy,
+  DEFAULT_STRATEGY_NAME,
+  StrategySelection,
+} from './strategy';
+
+export interface CdAware<U> extends Subscribable<U> {
+  nextValue: (value: any) => void;
+  nextConfig: (config: string) => void;
+}
+
+/**
+ * class CdAware
+ *
+ * @description
+ * This abstract class holds all the shared logic for the push pipe and the let directive
+ * responsible for change detection
+ * If you extend this class you need to implement how the update of the rendered value happens.
+ * Also custom behaviour is something you need to implement in the extending class
+ */
+export function createCdAware<U>(cfg: {
+  strategies: StrategySelection<U>;
+  resetContextObserver: NextObserver<any>;
+  updateViewContextObserver: PartialObserver<U> & NextObserver<U>;
+}): CdAware<U | undefined | null> {
+  const configSubject = new Subject<string>();
+  const config$: Observable<CdStrategy<U>> = configSubject.pipe(
+    distinctUntilChanged(),
+    startWith(DEFAULT_STRATEGY_NAME),
+    map(
+      (strategy: string): CdStrategy<U> =>
+        cfg.strategies[strategy]
+          ? cfg.strategies[strategy]
+          : cfg.strategies.idle
+    )
+  );
+
+  const observablesSubject = new Subject<Observable<U>>();
+  const observables$$ = observablesSubject.pipe(distinctUntilChanged());
+
+  let prevObservable: Observable<U>;
+  const renderSideEffect$ = combineLatest([observables$$, config$]).pipe(
+    switchMap(([observable$, strategy]) => {
+      if (prevObservable === observable$) {
+        return NEVER;
+      }
+
+      if (observable$ == null) {
+        cfg.updateViewContextObserver.next(observable$);
+        strategy.render();
+        return NEVER;
+      }
+
+      prevObservable = observable$;
+      cfg.resetContextObserver.next(observable$);
+      strategy.render();
+
+      return observable$.pipe(
+        distinctUntilChanged(),
+        tap(cfg.updateViewContextObserver),
+        strategy.behaviour(),
+        tap(() => strategy.render()),
+        catchError(e => EMPTY)
+      );
+    })
+  );
+
+  return {
```

Comment:
The curious thing about this API, it it could just be a `Subject<{ type: 'value', value: Observable<any> }| { type: 'config', value: config }>` or the like. I'd experiment with that and see if it ended up being cleaner or more useful.

Reply:
I don't understand how

How would I handle distinctUntilChange for observables and how can I provide both observable and config?

Reply:
I tried to get something working. No luck for me.